articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,421 @@
+In another example, the name `/foo` splits into one token, such that it comprises of a topic or service with the basename `foo` in the `/` namespace (the root namespace).
+
+Topic and service names:
+
+- may be specified as absolute or relative
+- must start with a forward slash (`/`) if they are absolute
+
+An absolute name begins with a forward slash (`/`) and does not respect namespaces, i.e. it can be considered "global".
+A relative name does not begin with a forward slash (`/`) and does respect the namespace of the node which created them.
+
+Relative names are appended to the namespace of the node which creates them.
+For example, if the node is put into a namespace `/ping/pong` and the topic or service name is `foo/bar`, then the absolute name will become `/ping/pong/foo/bar`.
+However, if the topic or service name is `/foo/bar`, then the absolute name will stay just `/foo/bar`, ignoring the node's namespace.
+
+### Name Tokens
+
+"Name token" are the strings between the namespace delimiters, e.g. the tokens of the topic or service `/foo/bar/baz` are `foo`, `bar`, and `baz`.
+
+Topic and service name tokens:
+
+- must not be empty, e.g. the name `//bar` is not allowed
+
+  - rational: it removes the chance for accidental `//` from concatenation
+
+- may use alphanumeric characters (`[0-9|a-z|A-Z]`), an underscore (`_`), or balanced curly braces (`{}`)
+
+- must not start with numeric characters (`[0-9]`)
+
+- must not end with a single underscore (`_`)
+
+  - rational: if tokens are allowed to end with and start with `_` then `foo_/bar` is indistinguishable from `foo/_bar` if the `/` is replaced with `__` (as proposed in the "Substitution of the Namespace Delimiter" section), i.e. both result in `foo___bar`
+
+- must not be a single underscore (`_`)
+
+  - rational: this is a special case of "must not end with an underscore"
+
+- must not have two or more underscores (`__`) repeated anywhere
+
+  - rational: this provides the implementation with a safe to use delimiter
+
+- may be a single tilde character (`~`)
+
+### Substitutions
+
+The bracket syntax (`{substitution_name}`) maybe used in non-fully qualified names to substitute useful contextual information into the name.
+The set of substitution keys (names) are not set in this document, but some reasonable examples might be: `{node}` expands to the current node's name or `{ns}` expands to the current node's namespace.
+
+Substitutions are expanded before the private namespace substitution character is expanded.
+Therefore a substitution may contain the private namespace substitution character, i.e. `~`.
+For example, given the name `{private}foo`, that there is a substitution called `{private}` which expands to `~/_`, the current node name is `my_node`, and the current node's namespace is `my_ns`, then the fully qualified name would be `/my_ns/my_node/_foo`.
```

Comment:
I'll update the document, if anyone thinks they should be nested they can comment here.